pkg/server/backend_manager.go

```diff
@@ -57,21 +60,66 @@ func newBackend(conn agent.AgentService_ConnectServer) *backend {
 	return &backend{conn: conn}
 }
 
-// BackendManager is an interface to manage backend connections, i.e.,
-// connection to the proxy agents.
-type BackendManager interface {
-	// Backend returns a single backend.
-	Backend() (Backend, error)
+// BackendStorage is an interface to manage the storage of the backend
+// connections, i.e., get, add and remove
+type BackendStorage interface {
+	// GetBackend returns a backend associating to the agentID.
+	GetBackend(agentID string) Backend
 	// AddBackend adds a backend.
 	AddBackend(agentID string, conn agent.AgentService_ConnectServer) Backend
 	// RemoveBackend removes a backend.
 	RemoveBackend(agentID string, conn agent.AgentService_ConnectServer)
+	// NumBackends returns the number of backends.
+	NumBackends() int
+}
+
+// BackendManager is an interface to manage backend connections, i.e.,
+// connection to the proxy agents.
+type BackendManager interface {
+	// Backend returns a single backend.
+	Backend(context.Context) (Backend, error)
+	BackendStorage
 }
 
 var _ BackendManager = &DefaultBackendManager{}
 
 // DefaultBackendManager is the default backend manager.
 type DefaultBackendManager struct {
+	*DefaultBackendStorage
+}
+
+func (dbm *DefaultBackendManager) Backend(_ context.Context) (Backend, error) {
+	be := dbm.DefaultBackendStorage.GetRandomBackend()
+	if be == nil {
+		return nil, &ErrNotFound{}
+	}
+	return be, nil
+}
+
+var _BackendManager = &DestIPBackendManager{}
+
+type DestIPBackendManager struct {
+	BackendStorage
+}
+
+func (dbm *DestIPBackendManager) Backend(ctx context.Context) (Backend, error) {
+	destIPInf := ctx.Value(DestIPContextKey)
+	if destIPInf == nil {
+		return nil, errors.New("no agentID found in the context")
```

Comment:
Not sure there is a legitimate way for the context to be missing this value, but if one crops up, I'm not sure that refusing to tunnel the traffic is the best available behavior.